Implementing maxAvailableComponentSets for resource models

mszacillo opened the pull request:
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

This PR implements the maxAvailableComponentsSets within general estimator based on resource models. The packing algorithm is first-fit decreasing, which is a greedy approach to bin packing. To make the calculation of the maxSets more efficient we use binary-search for checking the max set number, and then attempting to fit all the number of replicas onto the existing nodes. 

**Which issue(s) this PR fixes**:
Part of #6734

**Does this PR introduce a user-facing change?**:
```release-note
`karmada-scheduler`: Implements `getMaximumSetsBasedOnResourceModels` in general estimator.
```

mszacillo commented on pkg/estimator/client/general.go:
```diff
@@ -160,13 +160,197 @@ func (ge *GeneralEstimator) maxAvailableComponentSets(cluster *clusterv1alpha1.C
+	return true
+}
+
+func consumeAll(demand map[corev1.ResourceName]int64, cap map[corev1.ResourceName]int64) {
```
fixed!

zhzhuang-zju commented on pkg/estimator/client/general.go:
```diff
@@ -160,13 +160,197 @@ func (ge *GeneralEstimator) maxAvailableComponentSets(cluster *clusterv1alpha1.C
+			// Shouldn’t happen — status is malformed
+			return nil, fmt.Errorf("resource model/status mismatch: %d grades in spec, %d in status",
+				len(cluster.Spec.ResourceModels), len(cluster.Status.ResourceSummary.AllocatableModelings))
+		}
```
```suggestion
	if len(cluster.Spec.ResourceModels) != len(cluster.Status.ResourceSummary.AllocatableModelings) {
		// Shouldn’t happen — status is malformed
		return nil, fmt.Errorf("resource model/status mismatch: %d grades in spec, %d in status",
			len(cluster.Spec.ResourceModels), len(cluster.Status.ResourceSummary.AllocatableModelings))
	}
	for i := 0; i < len(cluster.Spec.ResourceModels); i++ {
```

Move the length check out of the for loop to avoid redundant evaluations.

zhzhuang-zju commented on pkg/estimator/client/general.go:
```diff
@@ -160,13 +160,197 @@ func (ge *GeneralEstimator) maxAvailableComponentSets(cluster *clusterv1alpha1.C
+			capCopy := make(map[corev1.ResourceName]int64, len(capTemplate))
+			for k, v := range capTemplate {
+				capCopy[k] = v
+			}
```
> This map copying logic is repeated in a few places in this file 

Yes, but compared to extracting it into a helper method, I prefer using `util.Resource`. First, it's a struct and doesn't require repeated copying; second, it encapsulates many resource-related operations, which can simplify the code.
https://github.com/karmada-io/karmada/blob/29536305bcc4c965a3538241bc60fb468d24d377/pkg/util/resource.go#L29-L37

mszacillo commented:
Thanks for the review @zhzhuang-zju !

> Does the "binary search" here actually improve efficiency?

Yes, binary search helps here. I implemented this method with the understanding that we are looking for an exact upperBound of possible component sets - rather than an approximation which would have been a lot simpler. 

We're looking for a max number of component sets between 0 and upperBound, with the knowledge that:

- If `S` sets are feasible, then all sets `s` <= `S` are also feasible
- If `S` sets are _not_ feasible, then all sets `s` >= `S` also also not feasible

We could naively loop through all possible set numbers [0, upperBound] and do the feasibility check, but given the rules listed above, binary search fits in quite well. And O(logN) < O(N).

> Since each iteration fully evaluates whether all upperBound component sets can be accommodated by the nodes, the total computation isn’t reduced—in fact, it might even increase due to repeated full validations.

Not sure I fully follow. By full validations you are referring to the feasibility fit check right? Even though each iteration fully validates feasibility, binary search reduces how many times the check is done. 

If we're discussing the feasibility check time complexity (assuming k = number of components), then we are sorting each time O(klogk), and then doing a greedy packing, which results in O(k * n * d), where n is nodes and d is the resources. Which given that number of components and resources is generally small, the complexity of this is generally going to depend on the number of nodes (larger clusters taking longer to check).  

That said! I'm realizing that we can move the sorting to be outside of the feasibility check, since the score we assign is dependent on the cluster. So it only needs to be computed once, and we can instead just set the replica count to match the # of sets we are running the feasibility check for. That would improve efficiency. 

> Given the high similarity between this logic and the noderesource plugin, can we unify their calculation logic and workflow?

I'm happy to unify calculation logic to make the code cleaner, but to me this sounds like a follow-up task. I could possibly even pick this up.

zhzhuang-zju commented:
> Not sure I fully follow. By full validations you are referring to the feasibility fit check right? Even though each iteration fully validates feasibility, binary search reduces how many times the check is done.

Binary search reduces the number of feasibility fit checks. However, is it possible to obtain the final result in a single feasibility fit check—for example, by assigning component sets one by one until we can no longer assign a complete component set?

> I'm happy to unify calculation logic to make the code cleaner, but to me this sounds like a follow-up task. I could possibly even pick this up.

Yes, you're absolutely right—this can be addressed as a follow-up task.

zhzhuang-zju commented on pkg/estimator/client/general.go:
```diff
@@ -160,13 +160,255 @@ func (ge *GeneralEstimator) maxAvailableComponentSets(cluster *clusterv1alpha1.C
+	oneSetKinds := expandKindsOneSet(components)
+	if len(oneSetKinds) == 0 {
+		// No pods in a set -> nothing to schedule under models
+		return 0, nil
```
The same as https://github.com/karmada-io/karmada/pull/6912#discussion_r2548795838

mszacillo commented on pkg/estimator/client/general.go:
```diff
@@ -160,13 +160,255 @@ func (ge *GeneralEstimator) maxAvailableComponentSets(cluster *clusterv1alpha1.C
+			limit = bound
+		}
+	}
+	if limit == math.MaxInt64 {
```
In this case no, because the helper method is used to determine maxFit of the component using the capacity of the cluster. We precheck cases in which the workloads have zero demand in the `maxAvailableComponentSets` method and return the podBound. Callers of `getMaximumSetsBasedOnResourceModels` will have resource requirements so it does not make sense to return maxInt here - that would throw off the calculation done by placeOneSet.

mszacillo commented:
> can you help rebase this PR? As we introduced some E2E tests regarding multiple components scheduling, so that we can see the test results from here.

Of course! Rebased and pushed the changes.

mszacillo commented:
@RainbowMango The recent changes to `buildModelNodes` made the order of the nodes nondeterministic. I'll fix this and update the branch.

mszacillo commented:
@RainbowMango I added a sort for the grades to make sure they are ordered when constructing the nodes. This could be optimized, but do you know if there is validation for grades users can define? Do they have to start from 0 and increment by 1? Or can users define random numbered grades if they wanted - for example {2, 6, 19, 22}?

RainbowMango commented on pkg/estimator/client/general.go:
```diff
@@ -160,13 +160,257 @@ func (ge *GeneralEstimator) maxAvailableComponentSets(cluster *clusterv1alpha1.C
+			capCopy := make(map[corev1.ResourceName]int64, len(tmpl))
+			for k, v := range tmpl {
+				capCopy[k] = v
+			}
```
```suggestion
			capCopy := maps.Clone(tmpl)
```
I just realized that now we have a built-in function for cloning a maps since [Go1.21](https://go.dev/doc/go1.21#maps).
